File: algorithms/lab3/heapsort/heapsort.py
from random import randint


def heapify(data, nums, root):
    largest = root
    left = 2 * root + 1
    right = 2 * root + 2

    if left < nums and data[root] < data[left]:
        largest = left

    if right < nums and data[largest] < data[right]:
        largest = right

    if largest != root:
        data[root], data[largest] = data[largest], data[root]
        heapify(data, nums, largest)


def heapsort(data):
    for root in range(len(data), -1, -1):
        heapify(data, len(data), root)

    for root in range(len(data) - 1, 0, -1):
        data[root], data[0] = data[0], data[root]  # swap
        heapify(data, root, 0)


if __name__ == "__main__":
    data = [randint(0, 100) for i in range(10)]
    print("Unsorted array is", data)
    heapsort(data)
    print("Sorted array is")
    for i in range(len(data)):
        print("%d" % data[i])

File: algorithms/lab3/heapsort/heapsort.js
function heapify(data, nums, root) {
    let largest = root;
    const left = 2 * root + 1;
    const right = 2 * root + 2;

    if (left < nums && data[root] < data[left]) {
        largest = left
    }

    if (right < nums && data[largest] < data[right]) {
        largest = right
    }

    if (largest !== root) {
        data[root] = data[largest];
        data[largest] = data[root];
        heapify(data, nums, largest)
    }
}

function heap_sort(data) {
    for (root = data.length; root > -1; root--) {
    heapify(data, data.length, root)
    }

    for (root = data.length - 1; root > 0; root--) {
        data[root] = data[0];
        data[0] = data[root];
        heapify(data, root, 0)
    }
}

const data = [5,7,4,2,8,6,1,3];
console.log('Unsorted array is', data);
heap_sort(data);
console.log('Sorted array is');
data.forEach(function (item, i, arr) {
   console.log(item)
});
